src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -503,6 +506,35 @@ void ObjectMonitor::enter_with_contention_mark(JavaThread *current, ObjectMonito
       // ParkEvent associated with this ObjectMonitor.
     }
 
+    ContinuationEntry* ce = current->last_continuation();
+    if (ce != nullptr && ce->is_virtual_thread()) {
+      int result = Continuation::try_preempt(current, ce->cont_oop(current));
+      if (result == freeze_ok) {
+        bool acquired = VThreadMonitorEnter(current);
+        if (acquired) {
+          // We actually acquired the monitor while trying to add the vthread to the
+          // _cxq so cancel preemption. We will still go through the preempt stub
+          // but instead of unmounting we will call thaw to continue execution.
+          current->set_preemption_cancelled(true);
+          if (JvmtiExport::should_post_monitor_contended_entered()) {
+            // We are going to call thaw again after this and finish the VMTS
+            // transition so no need to do it here. We will post the event there.
+            current->set_contended_entered_monitor(this);
+          }
+        }
+        current->set_current_pending_monitor(nullptr);
+        DEBUG_ONLY(int state = java_lang_VirtualThread::state(current->vthread()));
+        assert((acquired && current->preemption_cancelled() && state == java_lang_VirtualThread::RUNNING) ||
+               (!acquired && !current->preemption_cancelled() && state == java_lang_VirtualThread::BLOCKING), "invariant");
+        return;
+      }
+      if (result == freeze_pinned_native) {
+        post_virtual_thread_pinned_event(current, "Native frame or <clinit> on stack");
+      } else if (result == freeze_unsupported) {
+        post_virtual_thread_pinned_event(current, "Native frame or <clinit> or monitors on stack");
+      }
+    }
+
```

Comment:
Just a question. It is not clear from scratch why the `Continuation::try_preempt()` is called before the `VThreadMonitorEnter()`. It would be nice to add a comment explaining it. It can be also good to explain how it works together in this order. Even a surface explanation of a general idea would be very helpful. A part of this already explained in the comment at lines 515-517.